extensions/oauth/src/main/java/org/mule/extension/oauth2/internal/authorizationcode/DefaultAuthorizationCodeGrantType.java

```diff
@@ -247,53 +258,34 @@ private void buildHttpServer() throws InitialisationException {
 
   @Override
   public void authenticate(Event muleEvent, HttpRequestBuilder builder) throws MuleException {
-    final String resourceOwnerId = resourceOwnerIdEvaluator.resolveStringValue(muleEvent);
-    if (resourceOwnerId == null) {
-      throw new RequestAuthenticationException(createStaticMessage(String
-          .format("Evaluation of %s return an empty resourceOwnerId",
-                  localAuthorizationUrlResourceOwnerIdEvaluator.getRawValue())));
-    }
-    final String accessToken = getUserOAuthContext().getContextForResourceOwner(resourceOwnerId).getAccessToken();
+    final String accessToken =
+        getUserOAuthContext().getContextForResourceOwner(resolveExpression(resourceOwnerId, muleEvent))
+            .getAccessToken();
     if (accessToken == null) {
-      throw new RequestAuthenticationException(createStaticMessage(String.format(
-                                                                                 "No access token for the %s user. Verify that you have authenticated the user before trying to execute an operation to the API.",
-                                                                                 resourceOwnerId)));
+      throw new RequestAuthenticationException(createStaticMessage(format("No access token for the '%s' user. Verify that you have authenticated the user before trying to execute an operation to the API.",
+                                                                          resourceOwnerId.resolve())));
     }
     builder.addHeader(AUTHORIZATION, buildAuthorizationHeaderContent(accessToken));
   }
 
   @Override
   public boolean shouldRetry(final Event firstAttemptResponseEvent) throws MuleException {
-    if (!StringUtils.isBlank(getRefreshTokenWhen())) {
-      final Object value =
-          muleContext.getExpressionManager().evaluate(getRefreshTokenWhen(), firstAttemptResponseEvent).getValue();
-      if (!(value instanceof Boolean)) {
-        throw new MuleRuntimeException(createStaticMessage("Expression %s should return a boolean but return %s",
-                                                           getRefreshTokenWhen(), value));
-      }
-      Boolean shouldRetryRequest = (Boolean) value;
-      if (shouldRetryRequest) {
-        try {
-          refreshToken(firstAttemptResponseEvent, resourceOwnerIdEvaluator.resolveStringValue(firstAttemptResponseEvent));
-        } catch (MuleException e) {
-          throw new MuleRuntimeException(e);
-        }
+    Boolean shouldRetryRequest = resolveExpression(tokenRequestHandler.getRefreshTokenWhen(), firstAttemptResponseEvent);
```

Comment:
That is already being handled in AbstractAuthorizationCodeTokenRequestHandler#refreshToken